### ISSUE TITLE
TBW Widths

### ISSUE THREAD

**Issue by frostinassiky** [ekazakos/temporal-binding-network#12]
Hello Evangelos,

Thanks for publishing the code! They are well-organized and easy to follow ;)

One question here: when you ensemble five single TBN models, how do you set the different binding window widths?

**Comment by ekazakos** [comment]
Hi @Frostinassiky ,

Thanks for your comments and for using my code!

Well, for smaller TBW widths, instead of using 3 segments, Im using more. Doing so, each segment has a smaller width. Then, Im randomly sampling only 3 segments, by keeping the sparse sampling property.

Example for 6 segments:

(seg1 | seg2) | (seg3 | seg4) | (seg5 | seg6)

Parenthesis here show that I group every 2 segments, resulting in 3 groups. Then Im randomly sampling 1 segment from each group. This ensures sparse sampling, i.e. ensures that you will sample data covering the whole video.

**Comment by frostinassiky** [comment]
Hello @ekazakos 

Thanks for your quick response. 

If I understand the code correctly, you do not need to set a hyper-parameter for TBW width (_b_ in TBN paper). The TBW width is [average duration](https://github.com/ekazakos/temporal-binding-network/blob/4bc36e40c3c993c1f072e8341311ecfd86f9e42c/dataset.py#L114), the fraction of the number of frames over the number of segments. 

I have two more questions.

1. To re-implement the ensemble experiment, how many segments shall I set?

2. Also, I am curious why do you sample *3 segments* instead of 5, or more?

**Comment by ekazakos** [comment]
The TBW width would be the average of the "average duration" that you are mentioning over the entire training set. That is, because the TBW width is variable to the length of each clip, so the TBW widths that we are mentioning in the paper are averaged TBW widths over all the video clips in the training set, where the  TBW width for each clip is calculated using the "average duration". This approximates the TBW width used in an experiment.

1. I've used 3 different TBW widths for the experiment. If T is the duration of a clip, then the 3 TBW widths that were used to train different models were: T, T/3, T/6. T is the original one used in the code which is implemented by shuffling all the segments (found [here](https://github.com/ekazakos/temporal-binding-network/blob/4bc36e40c3c993c1f072e8341311ecfd86f9e42c/dataset.py#L166)). T/3 is implemented by removing the shuffling, so the TBW is only within segment and not across segments, and T/6 is implemented as described in my previous answer.

Now the problem is that I don't remember what exactly was the mix, i.e. how many models I trained for each width (since I have 5 models and the different widths are 3, more than one models were trained with the same width). But Im pretty sure that I trained at least 2 models with T.

2. Just for efficiency. The networks are already large (3 modalities x 3 segments = 9 nets in the memory).

If you have any further question please let me know.

**Comment by frostinassiky** [comment]
Thanks for your detailed explanation! I am happy to know that I don't need to re-implement the TBW width part by myself ;) The only missing implementation, the 6-segment experiment, was clearly described. 

In the ensemble, the possible mixes are [T, T, T/3, T/3, T/6] or  [T, T, T/3, T/6, T/6]. Hopefully, I can figure it out.

**Comment by TitaniumOne** [comment]
Thanks for your work and explanation! I have some questions:
1. I dont know the means of `In the TBN model instead, all modalities are trained simultaneously, and their combination is also learnt. ` Does it mean the variable b is learnable?  
2. I have no idea of how to implement the TBW width part, which is a learnable variable according to the paper.Have you delt with this problem? @ekazakos
